### PR TITLE
add traverseResult and sequenceResult for kotlin.Sequence

### PR DESCRIPTION
**Pull request by abendt**
https://github.com/arrow-kt/arrow/issues/2535

this adds _sequenceResult_ and _traverseResult_ for _kotlin.Sequence_

### CODE REVIEW

**Review by nomisRev — APPROVED** [review]
Thanks @abendt 👍

File: arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Sequence.kt
Comment (nomisRev) [inline arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/Sequence.kt:698]:
```suggestion
public inline fun <A, B> Sequence<A>.traverseResult(f: (A) -> Result<B>): Result<Sequence<B>> {
```

**Review by nomisRev — CHANGES_REQUESTED** [review]
Actually, sorry @abendt. That function should be `inline` to allow for the suspension to go through.

**Comment by abendt** [comment]
Another question @nomisRev. i also wanted to have a look at the _traverseX/sequenceX_ operators for Flows.
Is this a correct signature for e.g. the Result variant?

`public suspend fun <A, B> Flow<A>.traverseResult(f: suspend (A) -> Result<B>): Result<Flow<B>>`

**Comment by nomisRev** [comment]
Well the `Sequence` signature is actually already a bit missleading since it's actually processing **all** the values of the `Sequence`/`Flow` before returning. So it's actually not a sequencing/streaming operator, but a terminal operator.

```
public suspend fun <A, B> Sequence<A>.traverseResult(f: suspend (A) -> Result<B>): Result<List<B>>
public suspend fun <A, B> Flow<A>.traverseResult(f: suspend (A) -> Result<B>): Result<List<B>>
```

**Comment by abendt** [comment]
> Well the `Sequence` signature is actually already a bit missleading since it's actually processing **all** the values of the `Sequence`/`Flow` before returning. So it's actually not a sequencing/streaming operator, but a terminal operator.

true, that makes sense. Should i change the signature of the `Sequence.traverseResult` then too? Downside i see: there are the existing _traverseOption/Validated/..._  functions. These should ideally be changed too for consistency. Ofc this would be a breaking API change.

wdyt?

**Review by nomisRev — CHANGES_REQUESTED** [review]
Hey @abendt,

> rue, that makes sense. Should i change the signature of the Sequence.traverseResult then too?
> Ofc this would be a breaking API change.

I've been thinking about it, and I think we should actually deprecate the current `traverse` functions on `Sequence` since they're missleading. They're the equivalent of `sequence.toList().traverseXXX`.

So these functions actually promote an anti-pattern, so I would actually be in favor of deprecating them all together instead of updating them since an alias for `toList().traverse` seems redundant.

Since 1.0.0 we have to be very careful as to what functionality/methods we add since we can not easily remove/change the API. Having bloated, very large APIs surface is a downside not rather than an upside.

What do you think?

**Comment by nomisRev** [comment]
I looked into if we can make a sequence safe traverse function but since `SequenceScope` is designed as an `abstract class` with an `internal constructor` we cannot mix the effect as we could do if `SequenceScope` was an interface.
Otherwise we could create `interface EitherSequence<E, A> : EitherEffect<E>, SequenceScope<A>`.

**Comment by raulraja** [comment]
Thanks, @abendt. I propose we put this on hold, including adding additional functions to Arrow until the maintainers regroup in January. There is a plan to meet to discuss the Arrow Roadmap for 2022, including a review of needed methods and others that will get deprecated before 2.0.

**Pull request closed without merging**
